sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -202,3 +202,25 @@ object Unions {
     }
   }
 }
+
+/**
+ * A pattern that finds the original expression from a sequence of casts.
+ */
+object Casts {
+  def unapply(expr: Expression): Option[Attribute] = expr match {
+    case c: Cast => collectCasts(expr)
+    case _ => None
+  }
+
+  private def collectCasts(e: Expression): Option[Attribute] = {
+    if (e.isInstanceOf[Cast]) {
+      collectCasts(e.children(0))
+    } else {
+      if (e.isInstanceOf[Attribute]) {
+        Some(e.asInstanceOf[Attribute])
```

Comment:
Any time you see `isInstanceOf` / `asInstanceOf` you might want to be using matching instead...

```scala
e match {
  case a: Attribute => Some(a)
  case Cast(e, _) => Some(collectCasts(e))
  case _ => None
}
```